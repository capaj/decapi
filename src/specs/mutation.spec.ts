import { graphql } from 'graphql'
import {
  InputObjectType,
  InputField,
  ObjectType,
  Field,
  SchemaRoot,
  Mutation,
  compileSchema,
  Query
} from '../index'

@InputObjectType()
class BarType {
  @InputField()
  value: string
}

@ObjectType()
class Hello {
  constructor(private helloLabel = 'Hello') {}
  @Field()
  world(name: string): string {
    return `${this.helloLabel}, ${name}`
  }
}

@SchemaRoot()
class FooSchema {
  @Mutation()
  hello(): Hello {
    return new Hello()
  }

  @Query()
  deepInput(input: BarType): Hello {
    return new Hello(input.value)
  }
}

const schema = compileSchema(FooSchema)

describe('Mutation', () => {
  it('should not allow wrong argument types', async () => {
    const result = await graphql({
      schema,
      source: `
        mutation {
          hello {
            world(name: 2)
          }
        }
      `
    })
    expect(result.errors).toBeDefined()
    expect(result.errors).toMatchSnapshot()
  })

  it('should allow complex input types', async () => {
    const result = await graphql({
      schema,
      source: `
        {
          deepInput(input: { value: "Hola" }) {
            world(name: "Bob")
          }
        }
      `
    })

    expect(result).toEqual({ data: { deepInput: { world: 'Hola, Bob' } } })
  })
})
