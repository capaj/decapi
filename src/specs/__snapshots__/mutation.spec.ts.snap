// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation should not allow wrong argument types 1`] = `
Array [
  [GraphQLError: Expected type String!, found 2.],
]
`;
